torch/csrc/jit/passes/symbolic_shape_analysis.cpp

```diff
@@ -555,65 +526,230 @@ struct SymbolicShapeNodeAnalyzer {
+      GRAPH_DEBUG("Extracted Output: ", ss);
+      res.push_back(ss);
     }
+    return res;
   }
 
-  // node input indices that are TensorType and we need to iteratively
-  // substitute properties of. We only substitute properties
-  // of TensorTypes with a fixed dimension but not a complete shape,
-  // because a complete shape we can completely replace with a constant
-  // and non-fixed dimensions we cannot reason about at all
-  std::vector<std::pair<int64_t, ShapeArguments>> node_symbolic_input_indices_;
-  std::shared_ptr<Graph> shape_compute_graph_;
-  Node* node_;
+ public:
+  SymbolicShapeOpAnalyzer(const FunctionSchema* schema) : schema_(schema) {
+    shape_compute_graph_ = nullptr;
+    if (!schema_) {
+      return;
+    }
+    auto maybe_graph = shapeComputeGraphForSchema(*schema_);
+    if (!maybe_graph) {
+      return;
+    }
+    shape_compute_graph_ = (*maybe_graph)->copy();
+  }
+
+  c10::optional<std::vector<c10::SymbolicShape>> run(
+      std::vector<SsaArgument>& inputs) {
+    if (!shape_compute_graph_) {
+      return c10::nullopt;
+    }
+    inputs_ = inputs;
+    substituteConstantInputs();
+    GRAPH_DEBUG(inputs_)
+    return propagateShapesInGraph();
+  }
+
+  std::shared_ptr<Graph> getShapeComputeGraph() {
+    return shape_compute_graph_;
+  }
 };
 
+SsaArgument tensorShapeArg(Value* tensor_v) {
+  auto tt = tensor_v->type()->expect<TensorType>();
+  c10::SymbolicShape symbolic_shapes = tt->symbolic_sizes();
+
+  // for testing, we don't insert complete tensor shapes and rely on our
+  // partial evaluation pipeline to propagate information.
+  // this is a good proxy for our ability to propagate non-complete shape
+  // information.
+  if (symbolic_shapes.isComplete() && !symbolic_shape_analysis_test_mode) {
+    return IValue(tt->sizes().concrete_sizes());
+  }
+  if (toIValue(tensor_v)) {
+    auto size = constant_as<at::Tensor>(tensor_v)->sizes();
+    if (!symbolic_shape_analysis_test_mode) {
+      return IValue(size);
+    } else {
+      return c10::SymbolicShape(size);
+    }
+  }
+  return symbolic_shapes;
+}
+
+std::vector<SsaArgument> getNodeInputShapes(Node* n, const AliasDb& db) {
+  // TODO: fix the List of integers implementation, and
+  // extract out the shape changes, otherwise this is complete
+  Node* node_ = n;
+  // NB: shape compute graphs may have less inputs than their node
+  // counterparts to allow e.g. sharing one single unary definition
+  // so iterate on # of shape inputs
+  // We make lists of Tensor inputs variadic, which results in
+  // offset between a node index and its corresponding graph index
+  std::vector<SsaArgument> input_shapes = std::vector<SsaArgument>();
+
+  for (size_t node_index = 0; node_index < n->inputs().size(); ++node_index) {
+    auto type = node_->input(node_index)->type();
+
+    if (auto tt = type->castRaw<TensorType>()) {
+      input_shapes.push_back(tensorShapeArg(node_->input(node_index)));
+      continue;
```

Comment:
any particular reason to go from `if / else if/` to continue ? imo the former is a little clearer to read

Reply:
I find if()... continue easier to understand because I can know that there is no need to look further in the loop, whereas there can be additional logic hidden after a series of `if /else if` s